cmd/erasure-healfile.go

```diff
@@ -36,48 +50,86 @@ func (s ErasureStorage) HealFile(offlineDisks []StorageAPI, volume, path string,
 			f.Checksums[i] = checksums[i]
 		}
 	}
-	blocks := make([][]byte, len(s.disks))
+
+	// Scan part files in disk, block-by-block reconstruct it and
+	// write to stale disks.
 	chunksize := getChunkSize(blocksize, s.dataBlocks)
-	for offset := int64(0); offset < size; offset += blocksize {
-		if size < blocksize {
-			blocksize = size
+	var chunkOffset, blockOffset int64
+	for ; blockOffset < size; blockOffset += blocksize {
+		// last iteration may have less than blocksize data
+		// left, so chunksize needs to be recomputed.
+		if size < blockOffset+blocksize {
+			blocksize = size - blockOffset
 			chunksize = getChunkSize(blocksize, s.dataBlocks)
 		}
+
+		// read a chunk from each disk, until we have
+		// `s.dataBlocks` number of chunks set to non-nil in
+		// `blocks`
+		blocks := make([][]byte, len(s.disks))
+		var buffer []byte
 		numReads := 0
 		for i, disk := range s.disks {
-			if disk != OfflineDisk {
-				if blocks[i] == nil {
-					blocks[i] = make([]byte, chunksize)
-				}
-				blocks[i] = blocks[i][:chunksize]
-				if !verifiers[i].IsVerified() {
-					_, err = disk.ReadFileWithVerify(volume, path, offset, blocks[i], verifiers[i])
-				} else {
-					_, err = disk.ReadFile(volume, path, offset, blocks[i])
-				}
-				if err != nil {
-					blocks[i] = nil
-				} else {
-					numReads++
-				}
-				if numReads == s.dataBlocks { // we have enough data to reconstruct
-					break
-				}
+			if disk == OfflineDisk {
+				continue
+			}
+			// allocate buffer only when needed - when
+			// reads fail, the buffer can be reused
+			if int64(len(buffer)) != chunksize {
```

Comment:
This must be:
```
if  len(blocks[i]) == nil {
    blocks[i] = make([]byte, chunksize)
}
```

Reply:
Correction `len(blocks[i]) == 0`

Reply:
No, it is correct as written. The desired size of the buffer is how much we want to read. That is exactly `chunksize`. Remember that for last iteration chunksize may change - it is not a constant throughout the for-loop.

Reply:
Actually, I agree that @harshavardhana 's suggestion above will work as buffer is redeclared before the file-read loop. However, it is also correct as written, so I don't feel it necessary to change it.

Reply:
As said below it works - it is just not efficient in the sense of mem-allocs. On successful read buffer is always allocated. On `chunksize < len(buffer)` buffer is allocated. Instead of allocating the buffer on expected situations the buffer should be allocated on unexpected situations (read fails). Further if the`chunksize < len(buffer)` there is no reason to alloc a buffer. The buffer is large enough to hold to the data - it must only be truncated which is fine because the a smaller chunksize can only happen on last read.